File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-4.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        l = 0
        r = len(nums)-1

        while l <= r:
            m = l + (r-l)//2
            if m > 0 and nums[m] < nums[m-1]:
                return nums[m]
            if nums[0] <= nums[m]:
                l = m+1
            else:
                r = m-1

        return nums[0]

            

File: Data Structures & Algorithms/three-integer-sum/submission-4.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;
        int r, l, sum;
        sort(nums.begin(), nums.end());

        for (int i = 0; i < nums.size(); i++)
        {
            if (i > 0 and nums[i] == nums[i-1])
            {
                continue;
            }

            l = i+1;
            r = nums.size()-1;

            while (l < r)
            {
                sum = nums[i] + nums[l] + nums[r];
                if (sum > 0)
                {
                    r--;
                }
                else if (sum < 0)
                {
                    l++;
                }
                else
                {
                    res.push_back({nums[i], nums[l], nums[r]});
                    l++;
                    while (l < r and nums[l] == nums[l-1])
                    {
                        l++;
                    }
                }
            }
        }

        return res;
    }
};
